File: admin/src/api/analytics.ts
import http, { unwrap } from '@/api/http'

export type AnalyticsRange = 'day' | 'week' | 'month'

export interface AnalyticsOverviewKpis {
  activeTasks: string | number
  execRate: string
  apiRate: string
  points: string | number
  retention: string
  avgTasks: string
  pushRate: string
  activeUsers: string | number
  compare: string
  newUsers: string
}

export interface AnalyticsOverviewCharts {
  trend: {
    labels: string[]
    activeUsers: number[]
    execCounts: number[]
  }
  taskStatus: Array<{ name: string; value: number }>
  platformApi: { labels: string[]; values: number[] }
  topUsers: { labels: string[]; values: number[] }
}

export interface AnalyticsOverviewFunnel {
  steps: Array<{ label: string; value: number; note: string; loss: string }>
  losses: Array<{ title: string; count: string; desc: string }>
  conversionRate: string
  rangeLabel?: string
}

export interface AnalyticsOverviewData {
  range: AnalyticsRange
  kpis: AnalyticsOverviewKpis
  charts: AnalyticsOverviewCharts
  funnel: AnalyticsOverviewFunnel
  empty?: boolean
}

export async function fetchAnalyticsOverview(range: AnalyticsRange): Promise<AnalyticsOverviewData> {
  const res = await http.get<{ code: number; msg: string; data: AnalyticsOverviewData }>(
    '/api/admin/v1/analytics/overview',
    { params: { range } },
  )
  return unwrap(res)
}

export interface ExecRecord {
  execId: string
  taskId: string
  userId: string
  taskType: string
  startedAt: string
  endedAt: string
  durationMs: number
  result: string
  failReason: string
  points: number
  collectCount: number
}

export interface ExecRunsData {
  range: AnalyticsRange
  total: number
  success: number
  successRate: string
  avgDurationMs: number
  records: ExecRecord[]
}

export async function fetchExecRuns(range: AnalyticsRange): Promise<ExecRunsData> {
  const res = await http.get<{ code: number; msg: string; data: ExecRunsData }>(
    '/api/admin/v1/analytics/exec-runs',
    { params: { range } },
  )
  return unwrap(res)
}

export interface ApiCallRecord {
  requestId: string
  taskId: string
  execId: string
  platform: string
  calledAt: string
  result: string
  errorCode: string
  latencyMs: number
}

export interface PlatformStat {
  platform: string
  total: number
  success: number
  successRate: string
}

export interface ApiCallsData {
  range: AnalyticsRange
  total: number
  success: number
  successRate: string
  avgLatencyMs: number
  platformStats: PlatformStat[]
  records: ApiCallRecord[]
}

export async function fetchApiCalls(range: AnalyticsRange): Promise<ApiCallsData> {
  const res = await http.get<{ code: number; msg: string; data: ApiCallsData }>(
    '/api/admin/v1/analytics/api-calls',
    { params: { range } },
  )
  return unwrap(res)
}

export interface PushLogRecord {
  pushId: string
  taskId: string
  userId: string
  webhook: string
  sendAt: string
  sendResult: string
  callbackAt: string
  callbackResult: string
  newDataCount: number
  errorCode: string
  retryCount: number
}

export interface PushLogsData {
  range: AnalyticsRange
  total: number
  sendSuccess: number
  callbackSuccess: number
  deliveryRate: string
  notifyOnCount: number
  notifyOffCount: number
  records: PushLogRecord[]
}

export async function fetchPushLogs(range: AnalyticsRange): Promise<PushLogsData> {
  const res = await http.get<{ code: number; msg: string; data: PushLogsData }>(
    '/api/admin/v1/analytics/push-logs',
    { params: { range } },
  )
  return unwrap(res)
}

export interface AnalyticsUserRecord {
  userId: string
  feishuId: string
  phone: string
  deviceType: string
  pluginVersion: string
  remark: string
  firstUseAt: string
  lastActiveAt: string
  activeHours: string
  taskCount: number
  pointsConsumed: number
}

export interface UsersData {
  range: AnalyticsRange
  totalUsers: number
  activeUsers: number
  newUsers: number
  retention: string
  records: AnalyticsUserRecord[]
}

export interface UserDetailTask {
  taskId: string
  taskType: string
  keywordCount: number
  platforms: string[]
  status: string
  createdAt: string
}

export interface UserDetailPoint {
  consumeId: string
  taskId: string
  platform: string
  amount: number
  balance: number
  consumedAt: string
}

export interface UserDetail {
  userId: string
  feishuId: string
  phone: string
  deviceType: string
  pluginVersion: string
  remark: string
  firstUseAt: string
  lastActiveAt: string
  activeHours: string
  taskCount: number
  execCount: number
  execSuccessRate: string
  totalPoints: number
  tasks: UserDetailTask[]
  points: UserDetailPoint[]
}

export async function fetchUsers(range: AnalyticsRange): Promise<UsersData> {
  const res = await http.get<{ code: number; msg: string; data: UsersData }>(
    '/api/admin/v1/analytics/users',
    { params: { range } },
  )
  return unwrap(res)
}

export async function fetchUserDetail(userId: string): Promise<UserDetail> {
  const res = await http.get<{ code: number; msg: string; data: UserDetail }>(
    `/api/admin/v1/analytics/users/${userId}/detail`,
  )
  return unwrap(res)
}

export async function updateUserRemark(userId: string, remark: string): Promise<void> {
  await http.put(`/api/admin/v1/analytics/users/${userId}/remark`, { remark })
}

export interface AdminTaskRecord {
  id: string
  taskName: string
  status: string
  statusRaw: string
  action: string
  platform: string
  keywords: string[]
  userId: string
  intervalMinutes: number
  fetchCount: number
  successCount: number
  failedCount: number
  errorMessage: string
  taskStartTime: string
  taskEndTime: string
  nextRunAt: string
  createdAt: string
  updatedAt: string
}

export interface AdminTaskStats {
  total: number
  running: number
  stopped: number
  completed: number
}

export interface AdminTasksData {
  total: number
  page: number
  limit: number
  records: AdminTaskRecord[]
  stats: AdminTaskStats
}

export interface AdminTasksParams {
  page?: number
  limit?: number
  keyword?: string
  status?: string
  created_start?: string
  created_end?: string
}

export async function fetchAdminTasks(params: AdminTasksParams = {}): Promise<AdminTasksData> {
  const res = await http.get<{ code: number; msg: string; data: AdminTasksData }>(
    '/api/admin/v1/analytics/tasks',
    { params },
  )
  return unwrap(res)
}

File: server/admin/routers/analytics.py
"""管理端：运营看板聚合（埋点数据）。"""

from __future__ import annotations

from typing import Literal, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from admin.deps import get_db
from admin.schemas.response import admin_ok
from social_platform.services import analytics_service, task_service

router = APIRouter(prefix="/analytics", tags=["管理端-埋点看板"])


def _db_or_none(db: Session) -> Optional[Session]:
    if not task_service.database_configured():
        return None
    return db


@router.get("/overview")
def analytics_overview(
    range: Literal["day", "week", "month"] = Query(default="month"),
    db: Session = Depends(get_db),
):
    """数据概览 KPI + 图表 + 漏斗（Phase 1 基于已入库埋点聚合）。"""
    session = _db_or_none(db)
    if session is None:
        return admin_ok(data={"range": range, "kpis": {}, "charts": {}, "funnel": {}, "empty": True})
    data = analytics_service.get_overview(session, range)
    return admin_ok(data=data)


@router.get("/exec-runs")
def analytics_exec_runs(
    range: Literal["day", "week", "month"] = Query(default="day"),
    db: Session = Depends(get_db),
):
    """执行监控（Phase 2）。"""
    session = _db_or_none(db)
    if session is None:
        return admin_ok(data={"range": range, "total": 0, "success": 0, "successRate": "-", "avgDurationMs": 0, "records": []})
    data = analytics_service.get_exec_runs(session, range)
    return admin_ok(data=data)


@router.get("/api-calls")
def analytics_api_calls(
    range: Literal["day", "week", "month"] = Query(default="day"),
    db: Session = Depends(get_db),
):
    """API 监控（Phase 2）。"""
    session = _db_or_none(db)
    if session is None:
        return admin_ok(data={"range": range, "total": 0, "success": 0, "successRate": "-", "avgLatencyMs": 0, "platformStats": [], "records": []})
    data = analytics_service.get_api_calls(session, range)
    return admin_ok(data=data)


@router.get("/push-logs")
def analytics_push_logs(
    range: Literal["day", "week", "month"] = Query(default="day"),
    db: Session = Depends(get_db),
):
    """推送监控（Phase 3）。"""
    session = _db_or_none(db)
    if session is None:
        return admin_ok(data={"range": range, "total": 0, "sendSuccess": 0, "callbackSuccess": 0, "deliveryRate": "-", "notifyOnCount": 0, "notifyOffCount": 0, "records": []})
    data = analytics_service.get_push_logs(session, range)
    return admin_ok(data=data)


@router.get("/users")
def analytics_users(
    range: Literal["day", "week", "month"] = Query(default="month"),
    db: Session = Depends(get_db),
):
    """用户管理（Phase 3）。"""
    session = _db_or_none(db)
    if session is None:
        return admin_ok(data={"range": range, "totalUsers": 0, "activeUsers": 0, "newUsers": 0, "retention": "-", "records": []})
    data = analytics_service.get_users(session, range)
    return admin_ok(data=data)


@router.put("/users/{user_id}/remark")
def update_user_remark(
    user_id: str,
    body: dict,
    db: Session = Depends(get_db),
):
    """更新用户运营备注（Phase 3）。"""
    session = _db_or_none(db)
    if session is None:
        return admin_ok(data={"updated": False})
    remark = str(body.get("remark") or "")[:255]
    analytics_service.update_user_remark(session, user_id=user_id, remark=remark)
    return admin_ok(data={"updated": True})


@router.get("/users/{user_id}/detail")
def analytics_user_detail(
    user_id: str,
    db: Session = Depends(get_db),
):
    """用户详情（Phase 3）。"""
    session = _db_or_none(db)
    if session is None:
        return admin_ok(data=None)
    data = analytics_service.get_user_detail(session, user_id=user_id)
    return admin_ok(data=data)


@router.get("/tasks")
def analytics_tasks(
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=20, ge=1, le=100),
    keyword: Optional[str] = Query(default=None),
    status: Optional[str] = Query(default=None),
    created_start: Optional[str] = Query(default=None),
    created_end: Optional[str] = Query(default=None),
    db: Session = Depends(get_db),
):
    """任务管理（Phase 4）。"""
    session = _db_or_none(db)
    if session is None:
        return admin_ok(data={"total": 0, "page": page, "limit": limit, "records": [], "stats": {}})
    data = analytics_service.get_tasks(
        session,
        page=page,
        limit=limit,
        keyword=keyword,
        status=status,
        created_start=created_start,
        created_end=created_end,
    )
    return admin_ok(data=data)
